src/main/java/com/google/sps/servlets/PromotionQuizServlet.java

```diff
@@ -21,36 +26,44 @@
   private static final String JSON_CONTENT_TYPE = "application/json";
   private static final String QUIZ_SUBMIT = "promotionQuizSubmit";
   // this will be replaced by the game stage id queried from the user
-  private static final String HARDCODED_QUIZ_KEY = "softwareEngineerLevel1";
   private static final String PROMOTED_MESSAGE =
       "Congratulations, you passed the quiz and were promoted!";
   private static final String NOT_PROMOTED_MESSAGE =
       "You did not pass the quiz. Study the content and try again later";
-  private static final Double CORRECT_NEEDED_THRESHOLD = 0.5;
-
-  private static final Gson gson = new Gson();
+  private static final String IS_FINAL_STAGE_MESSAGE =
+      "Congratulations! You reached the final stage! You may no longer be promoted in this path.";
   // future: threshold based on level https://github.com/googleinterns/step144-2020/issues/89
-
-  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-  private QuestionDatabase questionDatabase;
+  private static final Double CORRECT_NEEDED_THRESHOLD = 0.5;
+  private static Gson gson;
+  private DatastoreService datastore;
+  private UserService userService;
+  private GameStageDatabase gameStageDatabase;
+  private PlayerDatabase playerDatabase;
   private ArrayList<QuizQuestion> quizQuestions;
 
   @Override
   public void init() {
-    String queryString = getGameStageSpecific_PromotionQuizQueryString();
-    this.setQuestionDatabase(new QuestionDatabase(datastore, queryString));
-  }
-
-  public void setQuestionDatabase(QuestionDatabase questionDatabase) {
-    this.questionDatabase = questionDatabase;
+    this.gson = new Gson();
+    this.userService = UserServiceFactory.getUserService();
+    this.datastore = DatastoreServiceFactory.getDatastoreService();
+    this.gameStageDatabase = new GameStageDatabase(datastore);
+    this.playerDatabase = new PlayerDatabase(datastore, userService);
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    this.quizQuestions = this.questionDatabase.getQuizQuestions();
-    String quizQuestionsJson = gson.toJson(quizQuestions);
     response.setContentType(JSON_CONTENT_TYPE);
-    response.getWriter().println(quizQuestionsJson);
+    try {
+      if (isUserOnFinalStage()) {
+        response.getWriter().println(gson.toJson(IS_FINAL_STAGE_MESSAGE));
+      } else {
+        this.quizQuestions = getQuizQuestions(response);
+        String quizQuestionsJson = gson.toJson(quizQuestions);
+        response.getWriter().println(quizQuestionsJson);
+      }
+    } catch (Exception e) {
```

Comment:
Here (and everywhere you catch Exception) we should catch a more specific question. i.e. If we get a NullPointerException (a bug in the code) we shouldn't treat that as a user not being logged in, we should just not handle it.

Reply:
So the only problem is that Leah's code throws an Exception (generic), whenever a user is logged out. So I am unsure as how to catch it more specifically.